### PR TITLE
CP-18173 add check to vm lifecycle: no migration when using nested virt

### PR DESCRIPTION
**Pull request by lindig**
The check is a little more exhaustive: Migration, suspension, and checkpointing must be blocked if either nested virtualisation is active or platform:nomigrate is true. This can be overridden with --force. The commit contains unit tests.

_This commit will be updated to use a library function that interprets boolean values from platform flags._

Signed-off-by: Christian Lindig <christian.lindig@citrix.com>

### CODE REVIEW

File: ocaml/xapi/xapi_vm_lifecycle.ml [hunk 1]
```diff
@@ -261,6 +261,32 @@ let check_protection_policy ~vmr ~op ~ref_str =
...
+ * [platform:nested-virt]. If we cannot find a boot record the VM can
+ * migrate. A VM can always migrate if strict=false.
+ **)
+
+let check_mobile strict vm =
+	let is_true str = (* will switch to function in xapi_vm_helpers.ml *)
+		str |> String.lowercase |> function
+		| "true"  | "1" -> true
+		| "false" | "0" -> false
+		| _ -> failwith (Printf.sprintf "can't parse %s as boolean" str)
```
Comment (andyhhp) [inline ocaml/xapi/xapi_vm_lifecycle.ml]: It is always better to use '%s' with quotes, which makes the error message easier to read if str contains a space or other non-alpha characters

**Comment by andyhhp** [comment]
Other than that minor formatting option, LGTM

**Comment by robhoes** [comment]
Minor, but I'd call the new file just "platform.ml", not "xapi_platform.ml". The "xapi_*.ml" files are mostly for the top-level XenAPI functions (by convention, though this is not followed consistently).

**Comment by lindig** [comment]
I picked `xapi_*` because it is living in the `xapi/` directory. In OCaml all files are living conceptually in the same name space - hence it is important to have unique names and if we want to migrate to Oasis it will be easier if we have no name clashes.

**Comment by robhoes** [comment]
@lindig If you are worried about conflicts, then perhaps vm_platform.ml (since it is about VM platform keys)?

**Comment by jonludlam** [comment]
Looks good to me. I'm happy with the `vm_platform.ml` suggestion.

**Comment by lindig** [comment]
I'll change the name to {{vm_platform.ml}}.

**Comment by lindig** [comment]
Re-based and changed the name to `vm_platform.ml`.

**Comment by robhoes** [comment]
I am going to revert this, because it broke VM resume/migration, and turned out not quite right anyway.

**Pull request merged**